detekt-rules-style/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/BracesOnWhenStatements.kt

```diff
@@ -0,0 +1,253 @@
+ *  when (a) {
+ *      1 ->
+ *          f1()
+ *      2 -> f2()
+ *  }
+ *  // singleLine = 'necessary'
+ *  when (a) {
+ *      1 -> f1()
+ *      2 -> f2()
+ *  }
+ *  // multiLine = 'necessary'
+ *  when (a) {
+ *      1 ->
+ *          f1()
+ *      2 -> {
+ *          f2()
+ *          f3()
+ *      }
+ *  }
+ *
+ *  // singleLine = 'consistent'
+ *  when (a) {
+ *      1 -> { f1() }
+ *      2 -> { f2() }
+ *  }
+ *  when (a) {
+ *      1 -> f1()
+ *      2 -> f2()
+ *  }
+ *  // multiLine = 'consistent'
+ *  when (a) {
+ *      1 -> {
+ *          f1()
+ *      }
+ *      2 -> {
+ *          f2()
+ *          f3()
+ *      }
+ *  }
+ *
+ *  // singleLine = 'always'
+ *  when (a) {
+ *      1 -> { f1() }
+ *      2 -> { f2() }
+ *  }
+ *  // multiLine = 'always'
+ *  when (a) {
+ *      1 -> {
+ *          f1()
+ *      }
+ *      2 -> {
+ *          f2()
+ *          f3()
+ *      }
+ *  }
+ *
+ *  </compliant>
+ */
+class BracesOnWhenStatements(config: Config = Config.empty) : Rule(config) {
+    override val issue = Issue(
+        javaClass.simpleName,
+        Severity.Style,
+        "Braces do not comply with the specified policy",
+        Debt.FIVE_MINS
+    )
+
+    @Configuration("single-line braces policy")
+    private val singleLine: BracePolicy by config("necessary") { BracePolicy.getValue(it) }
+
+    @Configuration("multi-line braces policy")
+    private val multiLine: BracePolicy by config("consistent") { BracePolicy.getValue(it) }
+
+    override fun visitWhenExpression(expression: KtWhenExpression) {
+        super.visitWhenExpression(expression)
+
+        val branches: List<KtWhenEntry> = walk(expression)
+        validate(branches, policy(expression))
+    }
+
+    private fun walk(expression: KtWhenExpression) = expression.entries
+
+    private fun validate(branches: List<KtWhenEntry>, policy: BracePolicy) {
+        val violators = when (policy) {
+            BracePolicy.Always -> {
+                branches.filter { !it.hasBraces() }
+            }
+
+            BracePolicy.Necessary -> {
+                branches.filter { !it.isMultiStatement() && it.hasBraces() }
+            }
+
+            BracePolicy.Never -> {
+                branches.filter { it.hasBraces() }
+            }
+
+            BracePolicy.Consistent -> {
+                val braces = branches.count { it.hasBraces() }
+                val noBraces = branches.count { !it.hasBraces() }
+                if (braces != 0 && noBraces != 0) {
+                    branches.take(1)
+                } else {
+                    emptyList()
+                }
+            }
+        }
+        violators.forEach { report(it, policy) }
+    }
+
+    private fun KtWhenEntry.hasBraces(): Boolean = expression is KtBlockExpression
+
+    private fun KtWhenEntry.parentCandidate(): PsiElement? = this.parent
+
+    private fun KtWhenEntry.isMultiStatement(): Boolean =
+        expression.let { (it is KtBlockExpression) && (it.statements.size > 1) }
```

Comment:
```suggestion
        (expression is KtBlockExpression) && (expression.statements.size > 1)
```
it's actually shorter without `let`, and also easier to understand. I would also remove the parens, there's no need for clarification of where the conditions are, because there's no weird precedence or complex expression.:
```suggestion
        expression is KtBlockExpression && expression.statements.size > 1
```

Reply:
But smart cast fails because of custom getter so we need `let` here. Or create a variable.

Reply:
Removed